trillian/ctdns/ctdns_server/main.go

```diff
@@ -0,0 +1,250 @@
+		if len(beMap) == 1 {
+			// If there's only one of them we use the blocking option as we can't
+			// serve anything until connected.
+			opts = append(opts, grpc.WithBlock())
+		}
+		conn, err := grpc.Dial(be.BackendSpec, opts...)
+		if err != nil {
+			glog.Exitf("Could not dial RPC udpServer: %v: %v", be, err)
+		}
+		defer conn.Close()
+		clientMap[be.Name] = trillian.NewTrillianLogClient(conn)
+	}
+
+	// Register DNS handlers for all the configured logs using the correct RPC
+	// client. Ignore any that don't specify a zone.
+	var zones int
+	for _, c := range cfg.LogConfigs.Config {
+		if len(c.DnsZone) > 0 {
+			zones++
+			if err := setupDNSHandler(clientMap[c.LogBackendName], *rpcDeadline, c); err != nil {
+				glog.Exitf("Failed to set up DNS log instance for %+v: %v", cfg, err)
+			}
+		}
+	}
+
+	if zones == 0 {
+		glog.Fatalf("No logs have a dns_zone configured. Exiting.")
+	}
+
+	// Handle metrics on the DefaultServeMux. We don't serve HTTP requests
+	// for clients, only DNS.
+	http.Handle("/metrics", promhttp.Handler())
+
+	// Return a 200 on the root, for GCE default health checking :/
+	http.HandleFunc("/", func(resp http.ResponseWriter, req *http.Request) { resp.WriteHeader(http.StatusOK) })
+
+	if *getSTHInterval > 0 {
+		// TODO(Martin2112): This is a hack to ensure the metrics etc. are created
+		// or the STH update will crash. Do something better.
+		ctfe.NewLogContext(0, "", ctfe.CertValidationOpts{}, nil, nil, ctfe.InstanceOptions{MetricFactory: prometheus.MetricFactory{}}, nil)
+		// Regularly update the internal STH for each log so our metrics stay up-to-date with any tree head
+		// changes.
+		for _, c := range cfg.LogConfigs.Config {
+			ticker := time.NewTicker(*getSTHInterval)
+			go func(c *configpb.LogConfig) {
+				glog.Infof("start internal get-sth operations on log %v (%d)", c.Prefix, c.LogId)
+				for t := range ticker.C {
+					glog.V(1).Infof("tick at %v: force internal get-sth for log %v (%d)", t, c.Prefix, c.LogId)
+					if _, err := ctfe.GetTreeHead(ctx, clientMap[c.LogBackendName], c.LogId, c.Prefix); err != nil {
+						glog.Warningf("failed to retrieve tree head for log %v (%d): %v", c.Prefix, c.LogId, err)
+					}
+				}
+			}(c)
+		}
+	}
+
+	// Bring up the DNS udpServer and serve until we get a signal not to.
+	go awaitSignal(func() {
+		os.Exit(1)
+	})
+	// TODO(Martin2112): Might need a separate metrics endpoint like CTFE.
+	// Bring up the UDP udpServer and allow time for it to start.
+	ch := make(chan bool)
+	udpServer := dns.Server{
+		Addr:       *metricsEndpoint,
+		Net:        "udp",
+		TsigSecret: nil,
+		NotifyStartedFunc: func() {
+			ch <- true
+		},
+	}
+
+	// Allow a short time for the UDP server to notify us that it is ready.
+	go udpServer.ListenAndServe()
+	select {
+	case res := <-ch:
+		glog.Infof("UDP udpServer has started OK: %v", res)
+	case <-time.After(*startupWait):
+		glog.Fatal("UDP udpServer not listening within timeout")
```

Comment:
`glog.Exitf` (no need for a stack trace)

Reply:
Done.